gc.c

```diff
@@ -3458,10 +3657,77 @@ vm_weak_table_gen_ivar_foreach_too_complex_replace_i(st_data_t *_key, st_data_t
     return iter_data->update_callback((VALUE *)value, iter_data->data);
 }
 
-struct st_table *rb_generic_ivtbl_get(void);
+struct st_table *rb_generic_fields_tbl_get(void);
+
+static int
+vm_weak_table_id_to_obj_foreach(st_data_t key, st_data_t value, st_data_t data)
+{
+    struct global_vm_table_foreach_data *iter_data = (struct global_vm_table_foreach_data *)data;
+
+    int ret = iter_data->callback((VALUE)value, iter_data->data);
+
+    switch (ret) {
+      case ST_CONTINUE:
+        return ret;
+
+      case ST_DELETE:
+        GC_ASSERT(rb_shape_obj_has_id((VALUE)value));
+        return ST_DELETE;
+
+      case ST_REPLACE: {
+        VALUE new_value = (VALUE)value;
+        ret = iter_data->update_callback(&new_value, iter_data->data);
+        if (value != new_value) {
+            DURING_GC_COULD_MALLOC_REGION_START();
+            {
+                st_insert(id_to_obj_tbl, key, (st_data_t)new_value);
```

Comment:
Why do we need to do an `st_insert` here? If we only change the value, then we can just use `st_foreach_with_replace` and directly update the value.

Reply:
It's indeed a missed opportunity, I'll see about changing that.